pyvista/plotting/_plotting.py

```diff
@@ -78,14 +78,13 @@ def prepare_smooth_shading(mesh, scalars, texture, split_sharp_edges, feature_an
 
     if split_sharp_edges:
         if is_polydata:
-            if has_scalars:
-                # we must track the original IDs with our own array
+            if has_scalars and use_points:
+                # we must track the original IDs with our own array since point
+                # indices may have changed
                 indices_array = '__orig_ids__'
-                if use_points:
-                    arr_sz = mesh.n_points
-                else:
-                    arr_sz = mesh.n_cells
+                arr_sz = mesh.n_points
                 mesh.point_data[indices_array] = np.arange(arr_sz, dtype=np.int32)
+
```

Comment:
If I understand correctly, the scheme is that
1. in case of `UnstructuredGrid` input we don't have to do this, because we already have the `'vtkOriginalPointIds'` array that we'll use later to restore the split scalars
2. in case of `PolyData` input with active cells we don't have to do this, because edge splitting is actually point splitting, so cells (and their scalars) are unchanged
3. so we only need to keep track of the original point indices in case the input was already a `PolyData` (hence no `'vtkOriginalPointIds'`) and it has active point scalars (hence merely splitting would mess with the scalars).

This raises the question, though: couldn't/shouldn't we do all this inside `compute_normals`, which is the method that actually changes the size of the mesh in case of a split? We'd have to handle every point array in the same way, but that would kind of make sense to me.

Anyway, if I understand the above correctly, this seems good to me.

Reply:
> 1. in case of `UnstructuredGrid` input we don't have to do this, because we already have the `'vtkOriginalPointIds'` array that we'll use later to restore the split scalars

Yes. More specifically, `extract_surface` on any dataset will return a `PolyData` containing the `'vtkOriginalPointIds'` array.



> 2. * in case of `PolyData` input with active cells we don't have to do this, because edge splitting is actually point splitting, so cells (and their scalars) are unchanged
> 3. * so we only need to keep track of the original point indices in case the input was already a `PolyData` (hence no `'vtkOriginalPointIds'`) and it has active point scalars (hence merely splitting would mess with the scalars).

Correct on both accounts.

> This raises the question, though: couldn't/shouldn't we do all this inside compute_normals, which is the method that actually changes the size of the mesh in case of a split?

That's not a bad proposal. In fact, there are cases in which I've done this in the past on personal projects. Implemented in:
915230a799755f8138cc2fbb04811d6b2877c1ef

> We'd have to handle every point array in the same way, but that would kind of make sense to me.

I'd argue that we only have to this for our new ID tracking array as this is already being done internally by VTK.

Reply:
> I'd argue that we only have to this for our new ID tracking array as this is already being done internally by VTK.

So the remaining question is why we have to do this whole tracking dance in the first place. And I suspect the answer is because the scalars might be passed as an explicit array. Otherwise (i.e. when the scalars come from point data) VTK would be handling it correctly anyway. But I can see how making use of that would not be feasible with the current implementation.

Reply:
> I suspect the answer is because the scalars might be passed as an explicit array.

Correct. In other projects using `pyvista`, I've had to generate the ID array separately, so I think any work here will also be used outside `_plotter.py`.

Reply:
OK, thanks for clearing that up. I'll review the updated state tomorrow.